corehq/apps/smsforms/models.py

```diff
@@ -237,3 +245,101 @@ def move_to_next_action(self):
+                )
+                return True
+            else:
+                return False
+
+    @classmethod
+    def could_maybe_claim_channel_for_session(cls, session):
+        """
+        Check if there's another session running on a channel, for heuristic purposes
+
+        Does not actually claim the channel, or guarantee that the session could.
+        A subsequent call to claim_channel_for_session could still return False.
+        """
+        running_session_info = cls.get_running_session_info_for_channel(session.get_channel())
+        return not running_session_info.session_id or running_session_info.session_id == session.session_id
+
+    @classmethod
+    def release_channel_for_session(cls, session):
+        """
+        This allows another session to claim the channel
+
+        The contact_id remains set, and incoming SMS will keep affinity with that contact
+        until another session claims the channel.
+        """
+        channel = session.get_channel()
+        with cls._critical_section(channel):
+            running_session_info = cls.get_running_session_info_for_channel(channel)
+            if cls.could_maybe_claim_channel_for_session(session):
+                # Drop the session_id but keep the contact_id
+                # This will let incoming SMS keep affinity with that contact_id until a new session starts
+                running_session_info = running_session_info._replace(session_id=None)
+                cls._set_running_session_info_for_channel(
+                    channel, running_session_info,
+                    # Keep affinity for 30 days
+                    30 * 24 * 60 * 60
+                )
+
+    @classmethod
+    def get_running_session_info_for_channel(cls, channel):
+        """
+        Returns RunningSessionInfo(session_id, contact_id) for the session currently claiming the channel
+        """
+        key = cls._channel_affinity_cache_key(channel)
+        return cache.get(key) or RunningSessionInfo(None, None)
+
+    @classmethod
+    def _set_running_session_info_for_channel(cls, channel, running_session_info, expiry):
+        key = cls._channel_affinity_cache_key(channel)
+        cache.set(key, running_session_info, expiry)
+
+    @staticmethod
+    def _channel_affinity_cache_key(channel):
+        return f'XFormsSessionSynchronization.value.{channel.backend_id}/{channel.phone_number}'
+
+    @staticmethod
+    def _critical_section(channel):
+        return CriticalSection([
+            f'XFormsSessionSynchronization.critical_section.{channel.backend_id}/{channel.phone_number}'
+        ], timeout=5 * 60)
+
+
+@quickcache(['contact_id', 'phone_number'])
+def get_channel_for_contact(contact_id, phone_number):
+    return Channel(
+        backend_id=PhoneNumber.get_phone_number_for_owner(contact_id, phone_number).backend.couch_id,
+        phone_number=phone_number,
+    )
+
+
+RunningSessionInfo = namedtuple('RunningSessionInfo', ['session_id', 'contact_id'])
```

Comment:
Could/should this also indicate contact type, since those can be either case ids or user ids? Or do you anticipate that it okay to assume these won't overlap, that users will always have some kind of globally unique id even when they're no longer in couch?

Reply:
Oh, that's a good point. I'm not sure how we want to handle that, but I just knew that for the time being we didn't need it

Reply:
I think I will ignore this avenue for now other than to look out for any place this might be an issue